blog/2019-09-ngrx-data-views/README.md

```diff
@@ -0,0 +1,419 @@
+
+The most effective way to handle this clustering in this example is to fetch all data for categories and tags with list calls when the application is starting up. This data can then be cached and referenced whenever needed. `@ngrx/entity` provides a perfect API for handling multiple entities in Angular applications. The data can be stored normalized exactly the way it is retrieved via the API.
+
+## Clustered data for Books
+
+Consider the following example:
+
+- We have entities for Authors:
+
+```ts
+export interface Author {
+  id: string;
+  firstName: string;
+  lastName: string;
+}
+```
+
+- We have entities for Tags:
+
+```ts
+export interface Tag {
+  id: string;
+  name: string;
+}
+```
+
+- We have entities for Books referencing authors and tags only via ID:
+
+```ts
+export interface Book {
+  id: string;
+  title: string;
+  authorIds: string[];
+  description: string;
+  tagIds: string[];
+}
+```
+
+With this we can easily set up `@ngrx/store` with three individual states using `@ngrx/entity`. Have a look at the [official documentation](https://ngrx.io/guide/entity) on how to do this.
+
+## Using parameterized selectors
+
+When implementing the book detail page for your application, you will most likely have multiple selectors for all the data needed to display a single book.
+When using [parameterized selectors](https://ngrx.io/guide/store/selectors#using-selectors-with-props), the book selector looks like this:
+
+```ts
+export const getBook = createSelector(
+  getBookEntities,
+  (books, bookId: string): Book => books[bookId]
+)
+```
+
+You can see here that we rely on the `getBookEntities` selectors as an input to this selector. `getBookEntities` was created with entity selectors provided by `@ngrx/entity` and it will give us an dictionary object containing all books. Our projector function then uses these entities and an additional parameter to look up the right book from the state. The selector can be used in the component like this:
+
+```ts
+book$ = this.store.pipe(select(getBook, this.bookId));
+```
+
+(The parameter `bookId` can come from the routing, be injected into the component or be a static reference. For this example it doesn't really matter. Have a look at the later following StackBlitz examples to see how we did it.)
+
+As the `getBook` selector only provides access to IDs of authors and tags, we have to compose additional selectors for the display of these entities. The one for the authors would look like this:
+
+```ts
+export const getAuthorsOfBook = createSelector(
+  getBook,
+  getAuthorEntities,
+  (book, authors): Author[] =>
+    book
+      ? book.authorIds.map(authorId => authors[authorId])
+      : []
+)
+```
+
+Have a look at how we used the previous selector `getBook` in the new selector. When using a parameterized selector in another selector, the new one also becomes a parameterized selector.
```

Comment:
This statement is interesting! I did not know this. Do you have any further references for this? For example, what happens if I use multiple parametrized selectors in composition?

Reply:
they get merged. The official documentation includes it with a simple counter example: https://ngrx.io/guide/store/selectors#using-selectors-with-props

Do we really want to get into details here?

Reply:
No, not too deep. Maybe provide further references for people who don't know.